### ISSUE TITLE
finetune 的target构建问题

### ISSUE THREAD

**Issue by fthjane** [FlagOpen/FlagEmbedding#129]
tinetune 的modeling.py 中的forward函数，target的构建，是根据scores 的size 生成的递增的序列么？这样的话，如何体现正样本比负样本的score高？是不是我哪里理解的有问题~~

    def forward(self, query: Dict[str, Tensor] = None, passage: Dict[str, Tensor] = None, teacher_score: Tensor = None):
        q_reps = self.encode(query)
        p_reps = self.encode(passage)

        if self.training:
            if self.negatives_cross_device:
                q_reps = self._dist_gather_tensor(q_reps)
                p_reps = self._dist_gather_tensor(p_reps)

            scores = self.compute_similarity(q_reps, p_reps)
            scores = scores / self.temperature
            scores = scores.view(q_reps.size(0), -1)
            target = torch.arange(scores.size(0), device=scores.device, dtype=torch.long)  #  <==== 这里 是创建一个与scores第一维大小相同的整数序列，从0开始递增，使用与scores相同的设备和数据类型吗？下面的那行代码是根据neg 的数量对target的值做扩展。然后就用 这个target 和 score 去计算交叉熵了。这样怎么能体现对比学习中的正样本比负样本的score高？
            target = target * (p_reps.size(0) // q_reps.size(0))
            loss = self.compute_loss(scores, target)

        else:
            scores = self.compute_similarity(q_reps, p_reps)
            loss = None
        return EncoderOutput(
            loss=loss,
            scores=scores,
            q_reps=q_reps,
            p_reps=p_reps,
        )

**Comment by staoxiao** [comment]
如果train_group_size=8, target就是[0, 8, 16,...]。
训练是时候一个query会计算和当前batch所有Passage的scores，对于第一个query来说，他的pos索引为0，第二个query的索引为8（前8个passage为第一个query的正样本和负样本)

**Comment by fthjane** [comment]
还想问下，有尝试过hingeloss么，CE改为hingeloss，是否有正向收益？

**Comment by fthjane** [comment]
> 还想问下，有尝试过hingeloss么，CE改为hingeloss，是否有正向收益？



> 如果train_group_size=8, target就是[0, 8, 16,...]。 训练是时候一个query会计算和当前batch所有Passage的scores，对于第一个query来说，他的pos索引为0，第二个query的pos索引为8（前8个passage为第一个query的正样本和负样本)

赞！ 解决了

**Comment by fthjane** [comment]
> > 还想问下，有尝试过hingeloss么，CE改为hingeloss，是否有正向收益？
> 
> > 如果train_group_size=8, target就是[0, 8, 16,...]。 训练是时候一个query会计算和当前batch所有Passage的scores，对于第一个query来说，他的pos索引为0，第二个query的pos索引为8（前8个passage为第一个query的正样本和负样本)
> 
> 赞！ 解决了

还想问下，有尝试过hingeloss么，CE改为hingeloss，是否有正向收益？

**Comment by fthjane** [comment]
> 如果train_group_size=8, target就是[0, 8, 16,...]。 训练是时候一个query会计算和当前batch所有Passage的scores，对于第一个query来说，他的pos索引为0，第二个query的pos索引为8（前8个passage为第一个query的正样本和负样本)

赞~  明白了！
还想问下，有尝试过hingeloss么，CE改为hingeloss，是否有正向收益？

**Comment by staoxiao** [comment]
CE在向量检索中最常用，还没有尝试过hingeloss。
